tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,266 @@
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned int) == 4,
+                "this function assumes an unsigned int is 32-bit");
+  unsigned int return_val = atomicExch(reinterpret_cast<unsigned int*>(dest),
+                                       reinterpret_cast<unsigned int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned long long int) == 8,
+                "this function assumes an unsigned long long int is 64-bit");
+  unsigned long long int return_val = atomicExch(reinterpret_cast<unsigned long long int*>(dest),
+                                       reinterpret_cast<unsigned long long int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicExch(reinterpret_cast<float*>(dest),
+                                       reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#ifndef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(double) == 8,
+                "this function assumes an double is 64-bit");
+  double return_val = atomicExch(reinterpret_cast<double*>(dest),
+                                       reinterpret_cast<double&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#endif
+
+#else
```

Comment:
Why is it advantageous to use CUDA atomics with NVHPC rather than use acc atomic capture unconditionally?

Reply:
CUDA atomics were used to 1) avoid weird compilation errors when compiling some OpenACC atomic constructs and 2) reduce the number of instructions (even though its performance benefit is unclear).

Reply:
Not sure I understand.  Is it a mix of things not compiling and places where you compared assembly code and saw benefit to use the CUDA atomics?  Could you elaborate?

Reply:
Using CUDA atomic intrinsics instead of the OpenACC atomic constructs is mainly due to compilation errors; the possible performance benefit is not verified.

Reply:
Now changed to use acc atomic capture.